File: Java/TowerOfHanoi.java
import java.util.Scanner;

public class TowerOfHanoi {
    public static void towerOfHanoi(int disks, char originRod, char destinyRod, char auxRod) {
        if (disks == 1) {
            // Base condition, when only one disk is left, we can freely move it to the destiny rod
            System.out.println("Move disk #1 from "
                    + originRod + " to " + destinyRod);
        } else {
            // If more than one disk remains, we divide the tower into smaller numbers of disks until one disk is left
            // First, we want to move all the disks from our origin to our aux rod
            towerOfHanoi(disks - 1, originRod, auxRod, destinyRod);
            // Then, we move the top disk of our origin rod to our destiny rod
            System.out.println("Move disk #" + disks + " from "
                    + originRod + " to " + destinyRod);
            // Finally we want tp move all disks from the aux rod to our destiny rod
            towerOfHanoi(disks - 1, auxRod, destinyRod, originRod);
        }
    }

    public static void main(String args[]) {
        Scanner s = new Scanner(System.in);
        System.out.print("Enter the number of disks: ");
        int n = s.nextInt();
        towerOfHanoi(n, 'A', 'C', 'B');
    }
}

/*
Example with 3 disks:

Input n = 3

Output:
Move disk #1 from A to C
Move disk #2 from A to B
Move disk #1 from C to B
Move disk #3 from A to C
Move disk #1 from B to A
Move disk #2 from B to C
Move disk #1 from A to C

TIME COMPLEXITY: O(a^n)
SPACE COMPLEXITY: O(n)

*/

File: Python/towerOfHanoi.py
def tower_of_hanoi(d, src, aux, trgt):
    if(d == 1):
        print('Move disk 1 from rod {} to rod {}.'.format(src, trgt))
        return
    # Recursion
    tower_of_hanoi(d - 1, src, trgt, aux)
    print('Move disk {} from rod {} to rod {}.'.format(d, src, trgt))
    tower_of_hanoi(d - 1, aux, src, trgt)

d = int(input('Enter the number of d: '))  
# src: A, aux: B, trgt: C
tower_of_hanoi(d, 'A', 'B', 'C')

'''
Input: 3
Output: Move disk 1 from rod A to rod C.
	Move disk 2 from rod A to rod B.
	Move disk 1 from rod C to rod B.
	Move disk 3 from rod A to rod C.
	Move disk 1 from rod B to rod A.
	Move disk 2 from rod B to rod C.
	Move disk 1 from rod A to rod C.
	
Time Complexity: O(2^n)
Space Complexity: O(n)
'''
